opengovsg/postmangovsg
# feat(backend): interruptible recipients list uploads

## Problem

Closes #1138 

## Solution

**Features**:
- Use [bee-queue](https://github.com/bee-queue/bee-queue) to handle processing/resuming recipients list uploads. This allows for interruption to recipient list processing (e.g. from server restart) to be detected and resumed on server restart.

**Improvements**:
- Refactored out recipient list processing logic to a common generic function that's reusable across all channels

## Tests
Run unit tests to ensure that `processUpload` works as expected

**Test resumption of upload**
- Start up backend, frontend and workers separately
- Manually introduce delay in the upload processing by patching the code [here](https://github.com/opengovsg/postmangovsg/blob/441874ef62bad293abcb2d93c31093031df33e73/backend/src/core/services/upload.service.ts#L319-L329)
```diff
try {
+ await new Promise((resolve) => setTimeout(resolve, 30000))
  await ParseCsvService.parseAndProcessCsv(
    downloadStream,
    onPreview(params),
    onChunk(params),
    onComplete({
      ...params,
      key: s3Key,
      filename,
    })
  )
```
- Through the frontend, create a campaign and upload a recipient list
- While upload is being processed, terminate the backend process
- Restart the backend process and observe the logs that the upload restarts as expected

## Deploy Notes

**New environment variables**:
- `UPLOAD_REDIS_URI` : URI to the redis database for recipient list upload job queue
- `UPLOAD_QUEUE_NAME` : Name of queue used to store upload jobs
- `UPLOAD_CONCURRENCY` : Maximum number of simultaneous active jobs
- `UPLOAD_CHECK_STALLED_INTERVAL` : How often to check for stalled jobs in milliseconds

**New dependencies**:
- `bee-queue`: Redis backed job queue


## TODOs
- [x] Backend tests for `processUpload`
- [x] ~Shift timeout mechanism to backend to ensure that overly long jobs are terminated and not left hogging resources.~ _Will tackle in a separate PR_
